test/framework/test_cache_method.py

```diff
@@ -0,0 +1,196 @@
+
+
+class TestCacheMethod(QiskitExperimentsTestCase):
+    """Test for cache_method decorator"""
+
+    def test_cache_args(self):
+        """Test cache_args=True"""
+
+        class CachedClass:
+            """Class with cached method"""
+
+            def __init__(self):
+                self.method_calls = 0
+
+            @cache_method(cache_args=True)
+            def method(self, *args, **kwargs):
+                """Test method for caching"""
+                self.method_calls += 1
+                return args, kwargs
+
+        obj = CachedClass()
+        size = 10
+        cached_vals = [obj.method(i, i) for i in range(size)]
+        for i, val in enumerate(cached_vals):
+            self.assertEqual(obj.method(i, i), val, msg="method didn't return cached value")
+        self.assertEqual(obj.method_calls, size, msg="Cached method was not evaluated once per arg")
+
+    def test_cache_args_kwargs(self):
+        """Test cache_args=True with args and kwargs"""
+
+        class CachedClass:
+            """Class with cached method"""
+
+            def __init__(self):
+                self.method_calls = 0
+
+            @cache_method(cache_args=True)
+            def method(self, *args, **kwargs):
+                """Test method for caching"""
+                self.method_calls += 1
+                return args, kwargs
+
+        obj = CachedClass()
+        args = (1, 2, 3)
+        names = ["a", "b", "c", "d"]
+        cached_vals = [obj.method(*args, name=name) for name in names]
+        for name, val in zip(names, cached_vals):
+            self.assertEqual(
+                obj.method(*args, name=name), val, msg="method didn't return cached value"
+            )
+        self.assertEqual(
+            obj.method_calls, len(names), msg="Cached method was not evaluated once per arg"
+        )
+
+    def test_cache_args_false(self):
+        """Test cache_args=False"""
+
+        class CachedClass:
+            """Class with cached method"""
+
+            def __init__(self):
+                self.method_calls = 0
+
+            @cache_method(cache_args=False)
+            def method(self, *args, **kwargs):
+                """Test method for caching"""
+                self.method_calls += 1
+                return args, kwargs
+
+        obj = CachedClass()
+        ret = obj.method(1999)
+        for i in range(10):
+            self.assertEqual(obj.method(i), ret, msg="method didn't return cached value")
+        self.assertEqual(obj.method_calls, 1, msg="Cached method was not evaluated once")
+
+    def test_require_hashable_raises(self):
+        """Test require_hashable=True"""
+
+        class CachedClass:
+            """Class with cached method"""
+
+            @cache_method(require_hashable=True)
+            def method(self, *args, **kwargs):
+                """Test method for caching"""
+                return args, kwargs
+
+        obj = CachedClass()
+        self.assertRaises(TypeError, obj.method, [1, 2, 3])
```

Comment:
Seems like this behavior is too strict because we usually use `Sequence` for type hint. Can we cast iterable to tuple when it's called with sequence?

Reply:
You cant use list for lru_cache, you have to make sure your arg is something hashable (like converting to tuple) before calling the cached function/method. The point of a caching decorator is to be very fast for checking matches, if you are doing inspection/formatting of the arg values it starts to defeats the point of using it since it adds a lot of overhead to the function call.

Reply:
So formatting must be performed on experiment side? User input values are often non-hashable form. I'm fine with either way.

Reply:
Yeah, this is same as LRU cache (you can see this sort of handling in tomo basis code where i explicitly have to convert sequence types to tuples before calling the cached functions).